propagators/awsxray/awsxray_propagator.go

```diff
@@ -0,0 +1,180 @@
+	sc := trace.SpanFromContext(ctx).SpanContext()
+	headers := []string{}
+	if !sc.TraceID.IsValid() || !sc.SpanID.IsValid() {
+		return
+	}
+	otTraceId := sc.TraceID.String()
+	xrayTraceId := traceIdVersion + traceIdDelimiter + otTraceId[0:traceIdFirstPartLength] +
+		traceIdDelimiter + otTraceId[traceIdFirstPartLength:]
+	parentId := sc.SpanID
+	samplingFlag := notSampled
+	if sc.TraceFlags == traceFlagSampled {
+		samplingFlag = isSampled
+	}
+
+	headers = append(headers, traceIdKey, kvDelimiter, xrayTraceId, traceHeaderDelimiter, parentIdKey,
+		kvDelimiter, parentId.String(), traceHeaderDelimiter, sampleFlagKey, kvDelimiter, samplingFlag)
+
+	carrier.Set(traceHeaderKey, strings.Join(headers, ""))
+}
+
+// Extract extracts a context from the carrier if it contains AWS X-Ray headers.
+func (awsxray AwsXray) Extract(ctx context.Context, carrier otel.TextMapCarrier) context.Context {
+	// extract tracing information
+	if h := carrier.Get(traceHeaderKey); h != "" {
+		sc, err := extract(h)
+		if err == nil && sc.IsValid() {
+			return trace.ContextWithRemoteSpanContext(ctx, sc)
+		}
+	}
+	return ctx
+}
+
+func extract(headerVal string) (trace.SpanContext, error) {
+	var (
+		sc             = trace.SpanContext{}
+		err            error
+		delimiterIndex int
+		part           string
+	)
+	pos := 0
+	for pos < len(headerVal) {
+		delimiterIndex = indexOf(headerVal, traceHeaderDelimiter, pos)
+		if delimiterIndex >= 0 {
+			part = headerVal[pos:delimiterIndex]
+			pos = delimiterIndex + 1
+		} else {
+			//last part
+			part = strings.TrimSpace(headerVal[pos:])
+			pos = len(headerVal)
+		}
+		equalsIndex := strings.Index(part, kvDelimiter)
+		if equalsIndex < 0 {
+			return empty, errInvalidTraceHeader
+		}
+		value := part[equalsIndex+1:]
+		if strings.HasPrefix(part, traceIdKey) {
+			sc.TraceID, err = parseTraceId(value)
+			if err != nil {
+				return empty, errMalformedTraceID
+			}
+		} else if strings.HasPrefix(part, parentIdKey) {
+			//extract parentId
+			sc.SpanID, err = trace.SpanIDFromHex(value)
+			if err != nil {
+				return empty, errInvalidSpanIDLength
+			}
+		} else if strings.HasPrefix(part, sampleFlagKey) {
+			//extract traceflag
+			sc.TraceFlags = parseTraceFlag(value)
+		}
+	}
+	return sc, nil
+}
+
+//returns position of the first occurence of a substring starting at pos index
+func indexOf(str string, substr string, pos int) int {
+	index := strings.Index(str[pos:], substr)
+	if index > -1 {
+		index += pos
+	}
+	return index
+}
+
+//returns trace Id if  valid else return invalid trace Id
+func parseTraceId(xrayTraceId string) (trace.ID, error) {
+	if len(xrayTraceId) != traceIdLength {
+		return empty.TraceID, errMalformedTraceID
+	}
+	if !strings.HasPrefix(xrayTraceId, traceIdVersion) {
+		return empty.TraceID, errMalformedTraceID
+	}
+
+	if xrayTraceId[traceIdDelimitterIndex1:traceIdDelimitterIndex1+1] != traceIdDelimiter ||
+		xrayTraceId[traceIdDelimitterIndex2:+traceIdDelimitterIndex2+1] != traceIdDelimiter {
```

Comment:
this looks odd to me `:+traceIdDelimitterIndex2` why is there a `+` in front of the var name?

Reply:
Also, I think you're always supposed to have spaces between variables and operators like `+`. 
So `traceIdDelimitterIndex1 + 1`

Reply:
the + was an accident, good catch - made the fixes